log4j2/src/main/java/com/google/common/flogger/backend/log4j2/ValueList.java

```diff
@@ -0,0 +1,156 @@
+import com.google.common.flogger.context.Tags;
+import org.checkerframework.checker.nullness.compatqual.NullableDecl;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/*
+ * A simple linked list designed to store multiple metadata values efficiently in a StringMap.
+ *
+ * <p>Metadata values in Flogger always have unique keys, but those keys can have the same label.
+ * Because Log4j2 uses a {@code String} keyed map, we would risk clashing of values if we just used
+ * the label to store each value directly. This class lets us store a list of values for a single label while
+ * being memory efficient in the common case where each label really does only have one value.
+ */
+final class ValueList implements Iterable<Object> {
+
+    private final Object value;
+    private final ValueList next; // Nullable
+
+    private static ValueList EMPTY = new ValueList(null, null);
+
+    ValueList(Object value, ValueList existingList) {
+        // caller ensures this is non null.
+        this.value = value;
+        // to ensure the iterator works
+        this.next = existingList == null ? EMPTY : existingList;
+    }
+
+    static Object maybeWrap(Object value, @NullableDecl Object existingValue) {
+        checkNotNull(value);
+        if (existingValue == null) {
+            return value;
+        } else {
+            // This should only rarely happen, so a few small allocations seems acceptable.
+            ValueList existingList = existingValue instanceof ValueList ? (ValueList) existingValue
+                    : new ValueList(existingValue, null);
+            return new ValueList(value, existingList);
+        }
+    }
+
+    @Override
+    public Iterator<Object> iterator() {
+        return new LinkedIterator(this);
+    }
+
+    static void appendValues(String label, Object valueOrList, MetadataKey.KeyValueHandler kvh) {
+        if (valueOrList instanceof ValueList) {
+            for (Object value : (ValueList) valueOrList) {
+                emit(label, value, kvh);
+            }
+        } else {
+            emit(label, valueOrList, kvh);
+        }
+    }
+
+    /**
+     * Tags are key-value mappings which cannot be modified or replaced. If you add
+     * the tag mapping "foo" -> true and later add the mapping "foo" -> false, you
+     * get the value "foo" mapped to both true and false. This is *very* deliberate
+     * since the key space for tags is global and the risk of two bits of code
+     * accidentally using the same tag name is real (e.g. you get "id=abcd" but you
+     * added "id=xyz" so you think this isn't your log entry, but someone else added
+     * "id=abcd" in a context you weren't aware of).
+     *
+     * Given three tag mappings: "baz" -> , "foo" -> true and "foo" -> false the
+     * Value list is going to flatten the tags in order to store them as tags=[baz,
+     * foo=false, foo=true].
```

Comment:
maybe stick the snippets in `<pre>{@code ... }</pre>` and put in a new paragraph to stop line wrapping.